### ISSUE TITLE
Seemingly incorrect gradient when nesting

### ISSUE THREAD

**Issue by MikeInnes** [pytorch/pytorch#32576]
## 🐛 Bug

The following code results in `0` when the expected answer is `-1`:

```python
def f1(x):
    def f2(y):
        return y*(x-y)
    return torch.autograd.grad(f2(x), x, create_graph=True)[0]

x = torch.tensor(1.0, requires_grad=True)
torch.autograd.grad(f1(x), x)[0]
```

## To Reproduce

Run the code above.

## Expected behavior

The code should result in `tensor(-1.)` rather than `tensor(0.)`.

## Environment

```
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.15.1
GCC version: Could not collect
CMake version: version 3.15.5

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.3
[pip3] torch==1.4.0
[pip3] torchvision==0.5.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.3                      199  
[conda] mkl_fft                   1.0.12           py37h5e564d8_0  
[conda] mkl_random                1.0.2            py37h27c97d8_0  
[conda] pytorch                   1.4.0                   py3.7_0    pytorch
[conda] torchvision               0.5.0                  py37_cpu    pytorch
```

## Additional context

The following code in JAX produces the expected behaviour, for comparison:

```python
def f1(x):
    def f2(y):
        return y*(x-y)
    return jax.jvp(f2, (x,), (1.0,))[1]

jax.jvp(f1, (1.0,), (1.0,))[1] # => DeviceArray(-1., dtype=float32)
```

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by albanD** [comment]
Hi,

When you evaluate `f2(x)`, you get `x*(x-x) = 0`.
So `f2` is the constant function that returns 0.
Its gradient is thus `0`.
And so `f1(.)` is just the constant function that always returns `0`.
So its gradient is `0` as well.

Could you explain why you expect to see `-1` here?

**Comment by MikeInnes** [comment]
Although `y` and `x` happen to be equal here, they are not the same variable, and in particular `x` should be held constant when calculating `df2/dy`. The derivative is `(f(x+ϵ)-f(x))/ϵ` with small `ϵ`, so we get

```julia
f(x+ϵ) = y*(x-y) = (x+ϵ)*(x - (x+ϵ)) = -(x+ϵ)*ϵ = -xϵ # ϵ^2 term is dropped
f(x+ϵ) - f(x) = -xϵ - 0 = -xϵ
(f(x+ϵ)-f(x))/ϵ = -x = -1
```

And the derivative of this with respect to `x` is of course `-1`. You can check this numerically with

```python
def D(f, x):
    ϵ = 0.01
    return (f(x+ϵ)-f(x))/ϵ

def f1(x):
    def f2(y):
        return y*(x-y)
    return D(f2, x)

D(f1, 1)
```

PyTorch also behaves consistently with this interpretation in a slightly-modified example (`f2` is still `0` for any `x`, but `df2/dy = -2x`):

```python
def f1(x):
    def f2(y):
        return y*(2*x-y)
    x2 = 2*x
    return torch.autograd.grad(f2(x2), x2, create_graph=True)[0]

x = torch.tensor(1.0, requires_grad=True)
torch.autograd.grad(f1(x), x)[0] # tensor(-2.)
```

**Comment by albanD** [comment]
I would argue this is expected behavior:
- if the `x` you implicitly capture and the one you use in `f2(x)` are the **same** Tensor (and same python object), then they are the same thing.
- If you want them to be considered as different objects, then you should make them different objects. For example by doing `x_prime = x.view_as(x)` and use it like `x2` in you example above.

Do you agree with this?

**Comment by MikeInnes** [comment]
I see what you mean. I guess this seems a bit hard to square with the mathematical notion of the gradient: this notation looks a lot like `d/dx f(x)` but actually computes something quite different. (Of course, mathematically `x′ = 1*x` or similar doesn't create a 'new object', so there's no possibility for the result to change based on an operation like that; and most ADs I've used follow that, hence the surprise.)

Perhaps it'd be useful to have a clear description of what `autograd.grad` is actually doing (if not quite calculating the mathematical gradient), since I haven't seen this clearly stated. Let me know if I've missed some description of gotchas like this in the docs though.

More generally, is there some independent way to confirm that PyTorch's gradients are correct by this definition? e.g. I might normally check against numerical diff or calculus, which won't work here. Or is it easier to consider cases like this as undefined behaviour?

**Comment by albanD** [comment]
Is the following a clearer explanation?
- Jax gradient API takes a function as input and differentiate that function.
- Pytorch gradient api takes the result of a program that ran and differentiate that program.

**Comment by ezyang** [comment]
@albanD With your JAX-style higher-order differentiation API, would we get the JAX result?

**Comment by albanD** [comment]
Given that this new API is defined as differentiating a function. This is the way I wanted to go.
But we can actually choose either way. The difference is simply to add a `inp = inp.detach.requires_grad_()` at the beginning if we want JAX's behavior, and nothing if we want the other behavior.

Do you agree that the JAX-like behavior would be better here?

**Comment by ezyang** [comment]
I think detaching is the right thing to do, for a functional style api.

**Comment by albanD** [comment]
Closing this issue as it is expected behavior.

**Comment by MikeInnes** [comment]
> Is the following a clearer explanation? ...

Sorry, that doesn't quite clear it up for me. Can you go into a bit more detail on how it's different to see my python code as a "function" vs as a "program" and how this affects gradients?

Is there any scope for adding more documentation on what `grad` is actually calculating?

**Comment by albanD** [comment]
The idea is that when you do `jax.jvp(f, (inp,), (1.,))`, you ask for the gradient of `f` wrt to it's input.

In pytorch, you don't have such api. You create a Tensor that tracks gradient and then use it for some computation and ask for gradient at the end for everything that happened before.
When you do `torch.autograd(out, inp)`, you ask for the gradients of everything that is involved in computing `out`. So if you capture other Tensors implicitly when computing `f(inp)`, they will be taken into account as pytorch does not differentiate `f` but the program that computes `out`.

This api allows you to do things like `model(inp).backward()` to populate all the gradients in your model without having to create a function that take all the parameters as input.


We could mention this in the autograd notes maybe to make it clearer for users that might stumble upon that edge case.

**Comment by MikeInnes** [comment]
Great, thanks. So just to be clear, results like the original one should be derived by considering the operator-overloading based implementation of autograd, bearing in mind what counts as the "same tensor" in the sense of object identity rather than mathematical variable or value? That seems justifiable but I certainly agree it's worth being clear about in the docs.

I'm also curious if PyTorch provides any kinds of guarantees about what operations will and won't return new tensors. For example, if I took a gradient w.r.t. `x2 = 1*x`, and PyTorch optimised this to return `x` directly rather than copying, that could change results in cases like this; and the same is true for functions from libraries built on PyTorch.

**Comment by albanD** [comment]
Yes there is a simple guarantee:
All out-of-place operation return a new Tensor (note that views, like `b = a.view(-1)` are out-of-place even though they share memery), all inplace operation (like `add_`, `unsqueeze_`, etc) return the same object :)
